routing/bicycle_directions.cpp

```diff
@@ -222,39 +213,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
+  adjacentEdges.m_outgoingTurns.candidates.reserve(outgoingEdges.size());
+  uniNodeId = UniNodeId(inEdge.GetFeatureId(), startSegId, endSegId, inEdge.IsForward());
+  m2::PointD const & ingoingPoint = inEdge.GetStartJunction().GetPoint();
+  m2::PointD const & junctionPoint = inEdge.GetEndJunction().GetPoint();
+
+  for (auto const & edge : outgoingEdges)
+  {
+    if (edge.IsFake())
+      continue;
+
+    auto const & outFeatureId = edge.GetFeatureId();
+    FeatureType ft;
+    if (!GetLoader(outFeatureId.m_mwmId).GetFeatureByIndex(outFeatureId.m_index, ft))
+      continue;
+
+    auto const highwayClass = ftypes::GetHighwayClass(ft);
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Error, ());
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Undefined, ());
+
+    double angle = 0;
+
+    if (inEdge.GetFeatureId().m_mwmId == edge.GetFeatureId().m_mwmId)
+    {
+      ASSERT_LESS(MercatorBounds::DistanceOnEarth(junctionPoint, edge.GetStartJunction().GetPoint()),
+                  turns::kFeaturesNearTurnMeters, ());
+      m2::PointD const & outgoingPoint = edge.GetEndJunction().GetPoint();
+      angle = my::RadToDeg(turns::PiMinusTwoVectorsAngle(junctionPoint, ingoingPoint, outgoingPoint));
+    }
+    else
+    {
+      // Note. In case of crossing mwm border
+      // (inEdge.GetFeatureId().m_mwmId != edge.GetFeatureId().m_mwmId)
+      // twins of inEdge.GetFeatureId() are considered as outgoing features.
+      // In this case that turn candidate angle is invalid and
+      // should not be used for turn generation.
+      adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+    }
+    adjacentEdges.m_outgoingTurns.candidates.emplace_back(angle, uniNodeId, highwayClass);
+  }
+}
+
+Segment BicycleDirectionsEngine::GetSegment(FeatureID const & featureId, uint32_t segId, bool forward) const
+{
+  CHECK(m_numMwmIds, ());
+
+  if (!featureId.m_mwmId.IsAlive())
+    MYTHROW(RoutingException, ("Mwm:", featureId.m_mwmId, "is not alive."));
+
+  NumMwmId const numMwmId =
+    m_numMwmIds->GetId(featureId.m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+  return Segment(numMwmId, featureId.m_index, segId, forward);
+}
+
+void BicycleDirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
+    RoadGraphBase const & graph, vector<Junction> const & path,
+    IRoadGraph::TEdgeVector const & routeEdges, my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_GREATER(pathSize, 1, ());
+  CHECK_EQUAL(routeEdges.size(), pathSize - 1, ());
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(UniNodeId(UniNodeId::Type::Mwm), AdjacentEdges(0)));
+  auto constexpr kInvalidSegId = std::numeric_limits<uint32_t>::max();
+  // |startSegId| is a value to keep start segment id of a new instance of LoadedPathSegment.
+  uint32_t startSegId = kInvalidSegId;
+  vector<Junction> prevJunctions;
+  vector<Segment> prevSegments;
+  for (size_t i = 1; i < pathSize; ++i)
```

Comment:
Roughly speaking if we do so we'll get something like that:
```c++
void BicycleDirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
    RoadGraphBase const & graph, vector<Junction> const & path,
    IRoadGraph::TEdgeVector const & routeEdges, my::Cancellable const & cancellable)
{
  size_t const pathSize = path.size();
  CHECK_GREATER(pathSize, 1, ());
  CHECK_EQUAL(routeEdges.size(), pathSize - 1, ());

  for (size_t i = 1; i < pathSize; ++i)
  {
    if (cancellable.IsCancelled())
      return;

    uint32_t startSegId = routeEdges[i - 1].GetSegId();

    IRoadGraph::TEdgeVector outgoingEdges;
    IRoadGraph::TEdgeVector ingoingEdges;
    vector<Junction> prevJunctions;
    vector<Segment> prevSegments;
    do
    {
      prevJunctions.push_back(path[i - 1]);
      if (m_generateTrafficSegs && !routeEdges[i - 1].IsFake())
      {
        prevSegments.push_back(GetSegment(routeEdges[i - 1].GetFeatureId(),
                                          routeEdges[i - 1].GetSegId(),
                                          routeEdges[i - 1].IsForward()));
      }
      GetEdges(graph, path[i], (i + 1 == pathSize), outgoingEdges, ingoingEdges);
    } while (!IsJoint(ingoingEdges, outgoingEdges, routeEdges[i - 1], routeEdges[i],
                      (i + 1 == pathSize), routeEdges[i - 1].GetFeatureId().IsValid()) && ++i);

    Edge const & inEdge = routeEdges[i - 1];
    uint32_t const inSegId = inEdge.GetSegId();
    CHECK_EQUAL(prevJunctions.size(), abs(static_cast<int32_t>(inSegId - startSegId)) + 1, ());

    prevJunctions.push_back(path[i]);

    AdjacentEdges adjacentEdges(ingoingEdges.size());
    UniNodeId uniNodeId(UniNodeId::Type::Mwm);
    GetUniNodeIdAndAdjacentEdges(outgoingEdges, inEdge, startSegId, inSegId + 1, uniNodeId,
                                 adjacentEdges.m_outgoingTurns);

    size_t const prevJunctionSize = prevJunctions.size();
    LoadedPathSegment pathSegment(UniNodeId::Type::Mwm);
    LoadPathAttributes(uniNodeId.GetFeature(), pathSegment);
    pathSegment.m_nodeId = uniNodeId;
    pathSegment.m_path = std::move(prevJunctions);
    // @TODO(bykoianko) |pathSegment.m_weight| should be filled here.

    // For LoadedPathSegment instances which contain fake feature id(s), |LoadedPathSegment::m_trafficSegs|
    // should be empty because there's no traffic information for fake features.
    // Traffic information is not supported for bicycle routes as well.
    // Method BicycleDirectionsEngine::GetSegment() returns false for fake features and for bicycle routes.
    // It leads to preventing pushing item to |prevSegments|. So if there's no enough items in |prevSegments|
    // |pathSegment.m_trafficSegs| should be empty.
    // Note. For the time being BicycleDirectionsEngine is used for turn generation for bicycle and car routes.
    if (m_generateTrafficSegs && prevSegments.size() + 1 == prevJunctionSize)
      pathSegment.m_trafficSegs = std::move(prevSegments);

    auto const it = m_adjacentEdges.insert(make_pair(uniNodeId, std::move(adjacentEdges)));
    ASSERT(it.second, ());
    UNUSED_VALUE(it);
    m_pathSegments.push_back(std::move(pathSegment));
  }
}
```
This refactoring gives us less amount of code but to prevent copypaste I had to remove important aliases: 
```
Junction const & prevJunction = path[i - 1];
Junction const & currJunction = path[i];
uint32_t startSegId = routeEdges[i - 1].GetSegId();
FeatureID const & inFeatureId = inEdge.GetFeatureId();
uint32_t const inSegId = inEdge.GetSegId();
bool const inIsForward = inEdge.IsForward();
```
and so on. In my opinion the code became less understandable. I propose to leave it as is.